File: desktopclient/src/main/java/com/goexp/galgame/gui/view/game/DateController.java
package com.goexp.galgame.gui.view.game;

import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.fxml.FXML;
import javafx.scene.control.ToggleButton;
import javafx.scene.control.ToggleGroup;
import javafx.scene.layout.TilePane;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.LocalDate;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class DateController {

    private final Logger logger = LoggerFactory.getLogger(DateController.class);


    public BooleanProperty onLoadProperty = new SimpleBooleanProperty(false);

    public BooleanProperty onYearLoadProperty = new SimpleBooleanProperty(false);

    public String title;

    public LocalDate from;

    public LocalDate to;

    @FXML
    private TilePane dateCon;

    @FXML
    private TilePane flowYear;

    private ToggleGroup yearSelect = new ToggleGroup();

    private ToggleGroup monthSelect = new ToggleGroup();

    @FXML
    private void initialize() {

        var monthNodes = IntStream.rangeClosed(1, 12).boxed()
                .map(month -> {
                    var tog = new ToggleButton();
                    tog.setUserData(month);

                    if (month.equals(LocalDate.now().getMonthValue()))
                        tog.setText("今月");
                    else
                        tog.setText(month + "月");

                    tog.setToggleGroup(monthSelect);

                    return tog;
                })
                .collect(Collectors.toList());

        dateCon.getChildren().setAll(monthNodes);


        monthSelect.selectedToggleProperty().addListener((o, old, newV) -> {
            if (newV != null) {

                var month = (Integer) newV.getUserData();

                from = LocalDate.now().withMonth(month).withDayOfMonth(1);
                to = from.plusMonths(1).minusDays(1);
                title = month + "月";

                onLoadProperty.set(true);

                logger.debug("Range:{}  {}", from, to);
                onLoadProperty.set(false);
            }

        });


        var yearNodes = IntStream.rangeClosed(2000, LocalDate.now().getYear()).boxed()
                .map(year -> {
                    var tog = new ToggleButton();
                    tog.setUserData(year);


                    if (year.equals(LocalDate.now().getYear()))
                        tog.setText("今年");
                    else
                        tog.setText(year + "年");

                    tog.setToggleGroup(yearSelect);

                    return tog;
                })
                .collect(Collectors.toList());

        flowYear.getChildren().setAll(yearNodes);

        yearSelect.selectedToggleProperty().addListener((o, old, newV) -> {
            if (newV != null) {

                var year = (Integer) newV.getUserData();

                from = LocalDate.of(year, 1, 1);
                to = LocalDate.of(year, 12, 31);
                title = year + "年";


                onYearLoadProperty.set(true);
                logger.debug("Range:{}  {}", from, to);
                onYearLoadProperty.set(false);
            }

        });
    }

    public void load() {
        onLoadProperty.set(false);
    }
}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/DateController.scala
package com.goexp.galgame.gui.view.game

import com.goexp.common.util.date.DateUtil
import com.goexp.galgame.gui.task.game.search.ByDateRange
import com.goexp.galgame.gui.util.TabManager
import com.goexp.galgame.gui.util.res.LocalRes
import com.goexp.galgame.gui.view.common.control.DataTab
import com.goexp.ui.javafx.DefaultController
import com.goexp.ui.javafx.control.cell.TextListCell
import com.mongodb.client.model.Filters.{and, gte, lte}
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.ListView
import scalafx.scene.image.ImageView
//import javafx.scene.image.ImageView

import scalafx.Includes._

import java.time.LocalDate
import scala.jdk.CollectionConverters._

class DateController extends DefaultController {
  @FXML private var flowMonth: ListView[Int] = _
  @FXML private var flowYear: ListView[Int] = _


  override protected def initComponent(): Unit = {

    /**
     * month
     */
    flowMonth.setCellFactory { _ =>
      TextListCell[Int] { month =>
        if (month == LocalDate.now.getMonthValue) "今月" else s"${month}月"
      }
    }

    val months = (1 to 12).asJava
    flowMonth.setItems(FXCollections.observableArrayList(months))


    /**
     * year
     */
    flowYear.setCellFactory { _ =>
      TextListCell[Int] { year =>
        if (year == LocalDate.now.getYear) "今年" else s"${year}年"
      }
    }

    val years = (2000 to LocalDate.now.getYear + 1).reverse.asJava
    flowYear.setItems(FXCollections.observableArrayList(years))
  }

  override protected def eventBinding(): Unit = {

    flowMonth.getSelectionModel().selectedItemProperty().addListener { (_, _, month) =>
      val from = LocalDate.now.withMonth(month).withDayOfMonth(1)
      val to = from.plusMonths(1).minusDays(1)
      val title = s"${month}月"

      logger.debug(s"Range:${from}  ${to}")

      val dateFilter = and(
        gte("publishDate", DateUtil.toDate(s"${from.toString} 00:00:00")),
        lte("publishDate", DateUtil.toDate(s"${to.toString} 23:59:59"))
      )

      TabManager().open(title, {
        new DataTab(ExplorerData(new ByDateRange(from, to),dateFilter)) {
          text = (title)
          graphic = new ImageView(LocalRes.DATE_16_PNG)
        }
      })
    }


    flowYear.getSelectionModel().selectedItemProperty().addListener { (_, _, year) =>
      val from = LocalDate.of(year, 1, 1)
      val to = LocalDate.of(year, 12, 31)
      val title = s"${year}年"

      logger.debug(s"Range:${from}  ${to}")

      val dateFilter = and(
        gte("publishDate", DateUtil.toDate(s"${from.toString} 00:00:00")),
        lte("publishDate", DateUtil.toDate(s"${to.toString} 23:59:59"))
      )


      TabManager().open(title, {
        new DataTab(ExplorerData(new ByDateRange(from, to),dateFilter)) {
          text = (title)
          graphic = new ImageView(LocalRes.DATE_16_PNG)
        }
      })
    }

  }
}